apiserver/charms.go

```diff
@@ -428,76 +430,41 @@ func (h *charmsHandler) processGet(r *http.Request, st *state.State) (string, st
 		return "", "", errors.Annotate(err, "cannot parse charm URL")
 	}
 
-	var filePath string
-	file := query.Get("file")
-	if file == "" {
-		filePath = ""
-	} else {
-		filePath = path.Clean(file)
-	}
-
-	// Prepare the bundle directories.
-	name := charm.Quote(curlString)
-	charmArchivePath := filepath.Join(
-		h.dataDir,
-		"charm-get-cache",
-		st.ModelUUID(),
-		name+".zip",
-	)
-
-	// Check if the charm archive is already in the cache.
-	if _, err := os.Stat(charmArchivePath); os.IsNotExist(err) {
-		// Download the charm archive and save it to the cache.
-		if err = h.downloadCharm(st, curl, charmArchivePath); err != nil {
-			return "", "", errors.Annotate(err, "unable to retrieve and save the charm")
-		}
-	} else if err != nil {
-		return "", "", errors.Annotate(err, "cannot access the charms cache")
+	fileArg := query.Get("file")
+	if fileArg != "" {
+		fileArg = path.Clean(fileArg)
 	}
-	return charmArchivePath, filePath, nil
-}
 
-// downloadCharm downloads the given charm name from the provider storage and
-// saves the corresponding zip archive to the given charmArchivePath.
-func (h *charmsHandler) downloadCharm(st *state.State, curl *charm.URL, charmArchivePath string) error {
+	// Ensure the working directory exists.
+	tmpDir := filepath.Join(h.dataDir, "charm-get-tmp")
```

Comment:
It would be great if we could stop using the temp-dir altogether. If we made some tweaks to the state/storage.Storage interface so that it returned an io.ReaderAt, we could do that.

mgo.GridFile does not implement io.ReaderAt, but it easily could from the looks of it. We could either add that upstream, or wrap the GridFile to seek&read (but would need to do so in a goroutine safe way).